### PR TITLE
use finalizers for NetworkAddonsConfig cleanup

### PR DESCRIPTION
**Pull request by phoracek**
Kubernetes garbage collector doesn't support setting namespaced
resource as an owner reference of a cluster-wide resource. Due to
that, we have to perform cleanup of NetworkAddonsConfig via
a finalizer.

### CODE REVIEW

**Comment by phoracek** [comment]
/cc @djzager @rthallisey 

Please review, it is urgent as a fix for https://bugzilla.redhat.com/show_bug.cgi?id=1712851. Thanks

File: pkg/controller/hyperconverged/hyperconverged_controller.go [hunk 1]
```diff
@@ -107,6 +118,27 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
 		return reconcile.Result{}, err
 	}
 
+	// Check if the instance was marked to be deleted, and if it is, perform cleanup
+	deleted := instance.GetDeletionTimestamp() != nil
```
Comment (djzager) [inline pkg/controller/hyperconverged/hyperconverged_controller.go]: This block looks similar to what I find in the [user-guide](https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#handle-cleanup-on-deletion) but for some reason this is hard for me to read. Can we simplify?

```
func Reconcile(...) (...) {
        ...
        if instance.ObjectMeta.DeletionTimestamp.IsZero() {
               // add the finalizer if it's not there                                                          
                if !contains(instance.ObjectMeta.Finalizers, FinalizerName) {                                  
                        instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, FinalizerName)       
                        return reconcile.Result{}, r.client.Update(context.TODO(), instance)                         
                } 
        } else {
                if contains(instance.ObjectMeta.Finalizers, FinalizerName) {                                   
                        // do the cleanup
                        // which can be it's own function if you wish, but for now only really
                        // removes the network-addons-config if certain conditions are met
                        ...
         
                        // remove the finalizer                            
                        instance.ObjectMeta.Finalizers = drop(instance.ObjectMeta.Finalizers, FinalizerName) 
                        return reconcile.Result{}, r.client.Update(context.TODO(), instance) 
                }                                         
        }
        ....
}
```

Comment (djzager) [reply pkg/controller/hyperconverged/hyperconverged_controller.go]: I still think that we should only be worrying about an `hyperconvergeds.hco.kubevirt.io` finalizer...not sure what requires the `foregroundDeletionFinalizer`.

Comment (djzager) [reply pkg/controller/hyperconverged/hyperconverged_controller.go]: I believe this would be easier to follow/maintain.

Comment (phoracek) [reply pkg/controller/hyperconverged/hyperconverged_controller.go]: If we don't remove foregroundDeletion, nobody will, it stucks. I can't say I understand where it comes from though.

File: pkg/controller/hyperconverged/hyperconverged_controller.go [hunk 2]
```diff
@@ -117,11 +149,11 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
 	}
 
 	// Create the KubeVirt ConfigMap if it doesn't already exist
-	result, err := manageComponentResource(kvConfig, "KubeVirtConfig", r.client)
+	err = manageComponentResource(kvConfig, "KubeVirtConfig", r.client)
```
[the remark this answers is not in the document]
Comment (phoracek) [reply pkg/controller/hyperconverged/hyperconverged_controller.go]: I added a code to the end of Reconcile function. Then this result was not really needed anymore. Also, I prefer not to expose edge API (Reconciliation result here, HTTP return codes in other projects) to anyone but the top level function. The function either fails or passes, but it should not decide whether we requeue, or not, or we do, but only after some time.

[the remark this answers is not in the document]
Comment (phoracek) [reply pkg/controller/hyperconverged/hyperconverged_controller.go]: (Will add it in a separate PR, was lazy)

File: pkg/controller/hyperconverged/hyperconverged_controller.go [hunk 3]
```diff
@@ -230,26 +257,82 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
...
 	// KWebUI failed to create, requeue
 	if err != nil {
-		return result, err
+		return reconcile.Result{}, err
 	}
-	return result, nil
+
+	// Everything went fine, automatically reconcile after after a minute without observed activity to
+	// make sure that even deployed objects without owner reference will be re-created if removed.
+	return reconcile.Result{RequeueAfter: time.Minute}, nil
```
[the remark this answers is not in the document]
Comment (phoracek) [reply pkg/controller/hyperconverged/hyperconverged_controller.go]: Understood and agreed.

File: pkg/controller/hyperconverged/hyperconverged_controller.go (hunk 1 above)
Comment (phoracek) [reply pkg/controller/hyperconverged/hyperconverged_controller.go]: Was trying to be granular with those, for case where we successfully clean-up only half of our CRs. However, that is duplicate of does-CR-still-exist check. I will change it to use a single finalizer.

**Comment by phoracek** [comment]
Hello @igoihman, your operator's resource is cluster-wide too, isn't it? Not sure how comes you don't hit the same issue as we do (maybe you do, but you don't know yet). Please check this PR and https://bugzilla.redhat.com/show_bug.cgi?id=1712851.

In short, if you set owner reference of cluster-wide (e.g. CDI) resource to namespaced (e.g. HyperConverged), bad stuff happens.

**Comment by ovirt-infra** [comment]
All tests passed

**Comment by phoracek** [comment]
@djzager your comments should be addressed. Could you please take another look?

**Comment by ovirt-infra** [comment]
All tests passed

**Comment by igoihman** [comment]
@phoracek we don't experience this issue, you can check cdi operator log.

**Comment by igoihman** [comment]
@phoracek I think it's a matter of missing operator RBAC permissions

**Comment by phoracek** [comment]
@igoihman yes, for us only network addons operator was failing on OKD with HCO. CDI was fine. Just wanted you to be aware of it if it ever happens.

Interesting, not sure how RBAC would help, but I can't say I understand Kubernetes garbage collector. Lemme check, thanks.

**Comment by igoihman** [comment]
Thanks @phoracek . keep me in the loop if you find something.

**Comment by phoracek** [comment]
@djzager I tested patched HCO, if recreated NetworkAddonsConfig when needed and it removed it during the cleanup.

**Pull request merged**